twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/QueryStatsHelper.java

```diff
@@ -0,0 +1,195 @@
+      operatorStats.add_input_cpu_millis = getMillisOrNegativeOne(obj.getString("addInputCpu"));
+      operatorStats.add_input_user_millis = getMillisOrNegativeOne(obj.getString("addInputUser"));
+      operatorStats.input_data_size_bytes = getBytesOrNegativeOne(obj.getString("inputDataSize"));
+      operatorStats.input_positions = obj.getJsonNumber("inputPositions").longValue();
+      operatorStats.sum_squared_input_positions = obj.getJsonNumber("sumSquaredInputPositions").doubleValue();
+      operatorStats.get_output_calls = obj.getJsonNumber("getOutputCalls").longValue();
+      operatorStats.get_output_wall_millis = getMillisOrNegativeOne(obj.getString("getOutputWall"));
+      operatorStats.get_output_cpu_millis = getMillisOrNegativeOne(obj.getString("getOutputCpu"));
+      operatorStats.get_output_user_millis = getMillisOrNegativeOne(obj.getString("getOutputUser"));
+      operatorStats.output_data_size_bytes = getBytesOrNegativeOne(obj.getString("outputDataSize"));
+      operatorStats.output_positions = obj.getJsonNumber("outputPositions").longValue();
+      operatorStats.blocked_wall_millis = getMillisOrNegativeOne(obj.getString("blockedWall"));
+      operatorStats.finish_calls = obj.getJsonNumber("finishCalls").longValue();
+      operatorStats.finish_wall_millis = getMillisOrNegativeOne(obj.getString("finishWall"));
+      operatorStats.finish_cpu_millis = getMillisOrNegativeOne(obj.getString("finishCpu"));
+      operatorStats.finish_user_millis = getMillisOrNegativeOne(obj.getString("finishUser"));
+      operatorStats.memory_reservation_bytes = getBytesOrNegativeOne(obj.getString("memoryReservation"));
+      operatorStats.system_memory_reservation_bytes = getBytesOrNegativeOne(obj.getString("systemMemoryReservation"));
+    } catch (Exception e) {
+      log.error(e, String.format("Error retrieving operator stats from JsonObject:\n%s\n", obj.toString()));
+      return null;
+    }
+
+    return operatorStats;
+  }
+
+  public static Map<int, QueryStageInfo> getQueryStages(QueryMetadata eventMetadata)
+  {
+    if (!eventMetadata.getPayload().isPresent())
+      return null;
+
+    String payload = eventMetadata.getPayload().get();
+    JsonReader jsonReader = Json.createReader(new StringReader(payload));
+    Queue<JsonObject> stageJsonObjs = new LinkedList<JsonObject>();
+    try {
+      stageJsonObjs.add(jsonReader.readObject());
+    } catch (Exception e) {
+      log.warn(e,
+        String.format("getQueryStages - Unable to extract JsonObject out of following blob:\n%s\n", payload));
+      return null;
+    }
+
+    Map<int, QueryStageInfo> stages = new HashMap<int, QueryStageInfo>();
+    while (!stageJsonObjs.isEmpty()) {
+      JsonObject cur = stageJsonObjs.poll();
+      String stageIdStr = "Unknown";
+      try {
+        stageIdStr = cur.getString("stageId");
+        int stageId = Integer.parseInt(stageIdStr.split(".")[1]);
+        stages.put(stageId, getQueryStageInfo(stageId, cur));
```

Comment:
This will put `null` as a value for some keys. IIRC, the scrooge generated java code will call the value's `write` method if the value type is another TBase class(QueryStageInfo). The error will happen at the serialization so we will still miss the whole record.